src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -473,6 +503,77 @@ public void RemoveAdditionalFile(string fullPath)
+        {
+            var extension = FileNameUtilities.GetExtension(dynamicFilePath)?.TrimStart('.');
+            if (extension?.Length == 0)
+            {
+                return;
+            }
+
+            foreach (var provider in _dynamicFileInfoProviders)
+            {
+                // skip unrelated providers
+                if (!provider.Metadata.Extensions.Any(e => string.Equals(e, extension, StringComparison.OrdinalIgnoreCase)))
+                {
+                    continue;
+                }
+
+                // don't get confused by _filePath and filePath
+                // VisualStudioProject._filePath points to csproj/vbproj of the project
+                // and the parameter filePath points to dynamic file such as cshtml and etc
+                // 
+                // also, provider is free-threaded. so find to call Wait rather than JTF
+                var fileInfo = provider.Value.GetDynamicFileInfoAsync(
+                    projectId: Id, projectFilePath: _filePath, filePath: dynamicFilePath, CancellationToken.None).WaitAndGetResult_CanCallOnBackground(CancellationToken.None);
+
+                if (fileInfo == null)
+                {
+                    continue;
+                }
+
+                fileInfo = FixUpDynamicFileInfo(fileInfo, dynamicFilePath);
+
+                // remember map between original dynamic file path to DynamicFileInfo.FilePath
+                _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
+                _sourceFiles.AddDynamicFile(provider.Value, fileInfo, folders);
+                return;
+            }
+        }
+
+        private DynamicFileInfo FixUpDynamicFileInfo(DynamicFileInfo fileInfo, string filePath)
+        {
+            // we might change contract and just throw here. but for now, we keep existing contract where one can return null for DynamicFileInfo.FilePath
+            if (string.IsNullOrEmpty(fileInfo.FilePath))
+            {
+                return new DynamicFileInfo(filePath, fileInfo.SourceCodeKind, fileInfo.TextLoader, fileInfo.DocumentServiceProvider);
+            }
+
+            return fileInfo;
+        }
+
+        public void RemoveDynamicSourceFile(string dynamicFilePath)
+        {
+            var provider = _sourceFiles.RemoveDynamicFile(GetDynamicFileInfoPath(dynamicFilePath));
+
+            // provider is free-threaded. so find to call Wait rather than JTF
+            provider.RemoveDynamicFileInfoAsync(
+                projectId: Id, projectFilePath: _filePath, filePath: dynamicFilePath, CancellationToken.None).Wait(CancellationToken.None);
+        }
+
+        private void OnDynamicFileInfoUpdated(object sender, string dynamicFilePath)
+        {
+            _sourceFiles.ProcessFileChange(dynamicFilePath, GetDynamicFileInfoPath(dynamicFilePath));
+        }
+
+        private string GetDynamicFileInfoPath(string dynamicFilePath)
+        {
+            return _dynamicFilePathMaps[dynamicFilePath];
```

Comment:
what happens if this table doesn't contain the file path?

Reply:
then, it throws.

Reply:
Can you switch to checking arguments and providing appropriate messages to callers using our entrypoints?  We should not be writing code that throws in a deeper component which then passes out to our callers.  The former indicates bugs in our callers, the latter indicates bugs in our own impl and we should not ship bugs like that.